src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,182 @@
+      VerbPhraseFlatFact(AstPrincipal prin, AstVerbPhrase verb_phrase)
+        : prin(std::move(prin)),
+          verb_phrase(std::move(verb_phrase)) {}
+    } VerbPhraseFlatFact;
+    using InternalFormType = std::variant<VerbPhraseFlatFact, AstPredicate>;
+    InternalFormType internal_form_;
+    AstFlatFact(InternalFormType internal_form)
+        : internal_form_(std::move(internal_form)) {}
+};
+
+// This is the AST node for facts which have two forms:
+//  - <flat_fact>
+//  - <principal> canSay <flat_fact>
+class AstFact {
+  public:
+    std::unique_ptr<AstFact> CreateFact(AstFlatFact flat_fact);
+    std::unique_ptr<AstFact> CreateFact(AstPrincipal prin, AstFact* fact);
+  private:
+    typedef struct CanSayFact {
+      AstPrincipal prin;
+      AstFact *fact;
+      CanSayFact(AstPrincipal prin, AstFact* fact)
+        : prin(std::move(prin)),
+          fact(fact) {}
+    } CanSayFact;
+    using InternalFormType = std::variant<AstFlatFact, CanSayFact>;
+    InternalFormType internal_form_;
+    AstFact(InternalFormType internal_form)
+      : internal_form_(std::move(internal_form)) {};
+};
+
+// AST Node for AstAssertion (this is a sub-expression of SaysAstAssertion) 
+// which has two forms:
+//  - <fact>                                      (unconditional)
+//  - <fact> :- <flat_fact_1> ... <flat_fact_n>   (conditional)
+class AstAssertion {
+  public:
+    std::unique_ptr<AstAssertion> CreateAssertion(AstFact fact);
+    std::unique_ptr<AstAssertion> CreateAssertion(AstFact lhs,
+        std::vector<AstFlatFact> rhs);
+  private:
+    typedef struct CondAssertion {
+      AstFact lhs;
+      std::vector<AstFlatFact> rhs;
+      CondAssertion(AstFact lhs, std::vector<AstFlatFact> rhs)
+        : lhs(std::move(lhs)), rhs(std::move(rhs)) {}
+    } CondAssertion;
+    using InternalFormType = std::variant<AstFact, CondAssertion>;
+    InternalFormType internal_form_;
+    AstAssertion(InternalFormType internal_form)
+      : internal_form_(std::move(internal_form)) {};
+};
+
+// AST node for saysAssertion, which prepends an assertion with:
+// "<principal> says"
+class AstSaysAssertion {
+  public:
+    AstSaysAssertion(AstPrincipal prin, std::vector<AstAssertion> assertions)
+      : prin_(std::move(prin)), assertions_(std::move(assertions)) {};
+  private:
+    AstPrincipal prin_;
+    std::vector<AstAssertion> assertions_;
+};
+
+// A query which can be used to test whether "<principal> says <fact>" is true
+// given the rules and base facts.
+class AstQuery {
+  public:
+    AstQuery(std::string name, AstPrincipal prin, AstFact fact)
+      : name_(std::move(name)),
+        prin_(std::move(prin)),
+        fact_(std::move(fact)) {}
+  private:
+    std::string name_;
+    AstPrincipal prin_;
+    AstFact fact_;
+};
+
+class AstProgram {
+  public:
+    AstProgram(std::vector<AstSaysAssertion> says_assertions,
+        std::vector<AstQuery> queries)
+      : says_assertions_(says_assertions),
+        queries_(queries) {}; 
```

Comment:
Use `std::move` for these initializations to allow copy elision.

Reply:
oops, fixed.